crates/sui-verifier/src/char_type_verifier.rs

```diff
@@ -0,0 +1,227 @@
+    // successful
+    if field_count != 1
+        || struct_def.field(0).unwrap().signature != TypeSignature(SignatureToken::Bool)
+    {
+        return Err(format!(
+            "characteristic type candidate {}::{} must have a single bool field only (or no fields)",
+            module.self_id(),
+            struct_name,
+        ));
+    }
+
+    if !struct_handle.type_parameters.is_empty() {
+        return Err(format!(
+            "characteristic type candidate {}::{} cannot have type parameters",
+            module.self_id(),
+            struct_name,
+        ));
+    }
+    Ok(())
+}
+
+/// Checks if this module's `init` function conformant with the characteristic type
+fn verify_init_function_char_type(
+    module: &CompiledModule,
+    fn_handle: &FunctionHandle,
+    struct_name: &str,
+    struct_handle: &StructHandle,
+) -> Result<(), String> {
+    let view = &BinaryIndexedView::Module(module);
+    let fn_sig = view.signature_at(fn_handle.parameters);
+    if fn_sig.len() != 2 || !is_char_type(view, &fn_sig.0[0], struct_handle) {
+        // check only the first parameter - the other one is checked in entry_points verification
+        // pass
+        return Err(format!(
+            "init function of a module containing characteristic type candidate must have {}::{} as the first parameter",
+            module.self_id(),
+            struct_name,
+        ));
+    }
+
+    Ok(())
+}
+
+// Checks if a given SignatureToken represents a characteristic type struct
+fn is_char_type(
+    view: &BinaryIndexedView,
+    tok: &SignatureToken,
+    struct_handle: &StructHandle,
+) -> bool {
+    if let SignatureToken::Struct(idx) = tok {
+        if view.struct_handle_at(*idx) == struct_handle {
+            return true;
+        }
+    }
+
+    false
+}
+
+/// Checks if this module's `init` function has a single parameter of TxContext type only
+fn verify_init_function_single_param(
+    module: &CompiledModule,
+    fn_handle: &FunctionHandle,
+) -> Result<(), String> {
+    let view = &BinaryIndexedView::Module(module);
+    let fn_sig = view.signature_at(fn_handle.parameters);
+    if fn_sig.len() != 1 {
+        return Err(format!(
+            "Expected exactly one parameter for {}::{}  of type &mut {}::{}::{}",
+            module.self_id(),
+            INIT_FN_NAME,
+            SUI_FRAMEWORK_ADDRESS,
+            TX_CONTEXT_MODULE_NAME,
+            TX_CONTEXT_STRUCT_NAME,
+        ));
+    }
+
+    Ok(())
+}
+
+/// Checks if this module function does not contain instantiation of the characteristic type
+fn verify_no_instantiations(
+    module: &CompiledModule,
+    fn_def: &FunctionDefinition,
+    struct_name: &str,
+    struct_def: &StructDefinition,
+) -> Result<(), String> {
+    let view = &BinaryIndexedView::Module(module);
+    if let Some(code) = &fn_def.code {
+        for bcode in &code.code {
+            if let Bytecode::Pack(struct_def_idx) = bcode {
```

Comment:
nit, I'd probably do one of these `let packed_struct_def_idx = match instr { Bytecode::Pack(idx) => idx, _ => continue };`

Could also do a `filter_map` if you prefer